File: mobile/modules/vibe-chat-native/ios/ChatEngineModule.swift
import ExpoModulesCore

public final class ChatEngineModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ChatEngine")

    Function("isSupported") {
      true
    }

    Function("setChatEngineConfig") { (payload: [String: Any]) in
      ChatEngine.shared.configure(payload)
    }

    Function("getChatEngineStatus") {
      ChatEngine.shared.getStatus()
    }

    Function("connectChatEngine") {
      ChatEngine.shared.connect()
    }

    Function("disconnectChatEngine") {
      ChatEngine.shared.disconnect()
    }

    Function("bindChatSurface") { (payload: [String: Any]) in
      ChatEngine.shared.bindSurface(payload)
    }

    Function("unbindChatSurface") { (payload: [String: Any]) in
      ChatEngine.shared.unbindSurface(payload)
    }

    Function("openChatChannel") { (payload: [String: Any]) in
      ChatEngine.shared.openChatChannel(payload)
    }

    Function("closeChatChannel") { (payload: [String: Any]) in
      ChatEngine.shared.closeChatChannel(payload)
    }

    Function("sendDeliveryReceipt") { (payload: [String: Any]) in
      ChatEngine.shared.sendDeliveryReceipt(payload)
    }

    Function("sendReadReceipt") { (payload: [String: Any]) in
      ChatEngine.shared.sendReadReceipt(payload)
    }

    Function("upsertLocalMessageStatus") { (payload: [String: Any]) in
      ChatEngine.shared.upsertLocalMessageStatus(payload)
    }

    Function("sendEncryptedMessage") { (payload: [String: Any]) in
      ChatEngine.shared.sendEncryptedMessage(payload)
    }

    Function("sendMessage") { (payload: [String: Any]) in
      ChatEngine.shared.sendMessage(payload)
    }

    Function("retryOutgoingMessage") { (payload: [String: Any]) in
      ChatEngine.shared.retryOutgoingMessage(payload)
    }

    Function("cancelOutgoingMessage") { (payload: [String: Any]) in
      ChatEngine.shared.cancelOutgoingMessage(payload)
    }

    Function("editMessage") { (payload: [String: Any]) in
      ChatEngine.shared.editMessage(payload)
    }

    Function("deleteMessage") { (payload: [String: Any]) in
      ChatEngine.shared.deleteMessage(payload)
    }

    Function("sendTypingState") { (payload: [String: Any]) in
      ChatEngine.shared.sendTypingState(payload)
    }

    Function("sendRecordingState") { (payload: [String: Any]) in
      ChatEngine.shared.sendRecordingState(payload)
    }

    Function("sendEditMessage") { (payload: [String: Any]) in
      ChatEngine.shared.sendEditMessage(payload)
    }

    Function("sendDeleteMessage") { (payload: [String: Any]) in
      ChatEngine.shared.sendDeleteMessage(payload)
    }

    Function("setChatMuted") { (payload: [String: Any]) in
      ChatEngine.shared.setChatMuted(payload)
    }

    Function("clearChat") { (payload: [String: Any]) in
      ChatEngine.shared.clearChat(payload)
    }

    Function("blockUser") { (payload: [String: Any]) in
      ChatEngine.shared.blockUser(payload)
    }

    Function("getChatProfileSummary") { (payload: [String: Any]) in
      ChatEngine.shared.getChatProfileSummary(payload)
    }

    Function("getPinnedMessages") { (payload: [String: Any]) in
      ChatEngine.shared.getPinnedMessages(payload)
    }

    Function("pinMessage") { (payload: [String: Any]) in
      ChatEngine.shared.pinMessage(payload)
    }

    Function("getChatJournal") {
      ChatEngine.shared.getJournal()
    }

    Function("clearChatJournal") {
      ChatEngine.shared.clearJournal()
    }

    Function("isTyping") { (payload: [String: Any]) in
      ChatEngine.shared.isTyping(payload)
    }

    AsyncFunction("fetchAgentConfig") { (chatId: String, promise: Promise) in
      ChatEngine.shared.fetchAgentConfig(chatId: chatId) { config in
        promise.resolve(config ?? [:])
      }
    }

    AsyncFunction("saveAgentConfig") { (chatId: String, config: [String: Any], promise: Promise) in
      ChatEngine.shared.saveAgentConfig(chatId: chatId, config: config) { success in
        promise.resolve(["success": success])
      }
    }

    AsyncFunction("deleteAgentConfig") { (chatId: String, promise: Promise) in
      ChatEngine.shared.deleteAgentConfig(chatId: chatId) { success in
        promise.resolve(["success": success])
      }
    }

    AsyncFunction("generateAgentPrompt") {
      (chatId: String, input: String, enabledTools: [String], promise: Promise) in
      ChatEngine.shared.generateAgentPrompt(
        chatId: chatId,
        input: input,
        enabledTools: enabledTools
      ) { payload in
        promise.resolve(payload ?? [:])
      }
    }

    // Shadow-mode bridge until native Phoenix transport is enabled.
    Function("setPresenceSnapshot") { (payload: [String: Any]) in
      let userIds = (payload["userIds"] as? [String]) ?? []
      _ = ChatEngine.shared.setPresenceSnapshot(userIds: userIds)
    }
  }
}

File: mobile/modules/vibe-chat-native/android/src/main/java/expo/modules/vibechatnative/ChatEngineModule.kt
package expo.modules.vibechatnative

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ChatEngineModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ChatEngine")

    Function("isSupported") {
      true
    }

    Function("setChatEngineConfig") { payload: Map<String, Any?> ->
      val context = appContext.reactContext ?: appContext.currentActivity?.applicationContext
      if (context == null) emptyMap<String, Any?>() else ChatEngine.configure(context, payload)
    }

    Function("getChatEngineStatus") {
      ChatEngine.getStatus()
    }

    Function("connectChatEngine") {
      ChatEngine.connect()
    }

    Function("disconnectChatEngine") {
      ChatEngine.disconnect()
    }

    Function("bindChatSurface") { payload: Map<String, Any?> ->
      ChatEngine.bindSurface(payload)
    }

    Function("unbindChatSurface") { payload: Map<String, Any?> ->
      ChatEngine.unbindSurface(payload)
    }

    Function("openChatChannel") { payload: Map<String, Any?> ->
      ChatEngine.openChatChannel(payload)
    }

    Function("closeChatChannel") { payload: Map<String, Any?> ->
      ChatEngine.closeChatChannel(payload)
    }

    Function("sendDeliveryReceipt") { payload: Map<String, Any?> ->
      ChatEngine.sendDeliveryReceipt(payload)
    }

    Function("sendReadReceipt") { payload: Map<String, Any?> ->
      ChatEngine.sendReadReceipt(payload)
    }

    Function("upsertLocalMessageStatus") { payload: Map<String, Any?> ->
      ChatEngine.upsertLocalMessageStatus(payload)
    }

    Function("sendEncryptedMessage") { payload: Map<String, Any?> ->
      ChatEngine.sendEncryptedMessage(payload)
    }

    Function("sendMessage") { payload: Map<String, Any?> ->
      ChatEngine.sendMessage(payload)
    }

    Function("retryOutgoingMessage") { payload: Map<String, Any?> ->
      ChatEngine.retryOutgoingMessage(payload)
    }

    Function("cancelOutgoingMessage") { payload: Map<String, Any?> ->
      ChatEngine.cancelOutgoingMessage(payload)
    }

    Function("editMessage") { payload: Map<String, Any?> ->
      ChatEngine.editMessage(payload)
    }

    Function("deleteMessage") { payload: Map<String, Any?> ->
      ChatEngine.deleteMessage(payload)
    }

    Function("sendTypingState") { payload: Map<String, Any?> ->
      ChatEngine.sendTypingState(payload)
    }

    Function("sendRecordingState") { payload: Map<String, Any?> ->
      ChatEngine.sendRecordingState(payload)
    }

    Function("sendEditMessage") { payload: Map<String, Any?> ->
      ChatEngine.sendEditMessage(payload)
    }

    Function("sendDeleteMessage") { payload: Map<String, Any?> ->
      ChatEngine.sendDeleteMessage(payload)
    }

    Function("setChatMuted") { payload: Map<String, Any?> ->
      ChatEngine.setChatMuted(payload)
    }

    Function("clearChat") { payload: Map<String, Any?> ->
      ChatEngine.clearChat(payload)
    }

    Function("blockUser") { payload: Map<String, Any?> ->
      ChatEngine.blockUser(payload)
    }

    Function("getChatProfileSummary") { payload: Map<String, Any?> ->
      ChatEngine.getChatProfileSummary(payload)
    }

    Function("getPinnedMessages") { payload: Map<String, Any?> ->
      ChatEngine.getPinnedMessages(payload)
    }

    Function("pinMessage") { payload: Map<String, Any?> ->
      ChatEngine.pinMessage(payload)
    }

    Function("getChatJournal") {
      ChatEngine.getJournal()
    }

    Function("clearChatJournal") {
      ChatEngine.clearJournal()
    }

    Function("isTyping") { payload: Map<String, Any?> ->
      ChatEngine.isTyping(payload)
    }

    AsyncFunction("fetchAgentConfig") { payload: Map<String, Any?> ->
      ChatEngine.fetchAgentConfig(payload)
    }

    AsyncFunction("saveAgentConfig") { payload: Map<String, Any?> ->
      ChatEngine.saveAgentConfig(payload)
    }

    AsyncFunction("deleteAgentConfig") { payload: Map<String, Any?> ->
      ChatEngine.deleteAgentConfig(payload)
    }

    AsyncFunction("generateAgentPrompt") { payload: Map<String, Any?> ->
      ChatEngine.generateAgentPrompt(payload)
    }

    // Shadow-mode bridge until native Phoenix transport is implemented.
    Function("setPresenceSnapshot") { payload: Map<String, Any?> ->
      val list = (payload["userIds"] as? List<*>)?.mapNotNull { it?.toString() } ?: emptyList()
      ChatEngine.setPresenceSnapshot(list)
    }
  }
}
